### PR TITLE
Make DecimalArray as PrimitiveArray

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Part of #2637.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by viirya** [comment]
Hopefully to move this forward.

cc @sunchao @tustvold

File: arrow-array/src/array/primitive_array.rs [hunk 1]
```diff
@@ -827,9 +831,178 @@ impl<T: ArrowPrimitiveType> From<ArrayData> for PrimitiveArray<T> {
...
+        Ok(unsafe { data.build_unchecked().into() })
+    }
+
+    /// Returns a Decimal array with the same data as self, with the
+    /// specified precision.
+    ///
+    /// # Safety
+    ///
+    /// This doesn't validate decimal values with specified precision.
+    pub unsafe fn unchecked_with_precision_and_scale(
```
Comment (tustvold) [inline arrow-array/src/array/primitive_array.rs:876]: The move to PrimitiveArray implicitly means away from validating the contents for overflow, as any value of the underlying primitive is permitted.

I think we should therefore make this safe, and remove any value validation except for when this method is explicitly called by the user, as an opt-in

**Review by tustvold — COMMENTED** [review]
Thank you for this, just got off a 26 hour flight so will review this more thoroughly once my brain is functioning again properly. Only major comment is regards to validation, which I think can be relaxed

**Comment by viirya** [comment]
These clippy errors are not from this change but due to recent change on `invalid_html_tags`, fixing it at #2862.

**Comment by alamb** [comment]
cc @liukun4515

**Comment by alamb** [comment]
If / when this PR is ready to merge, I would like to try to update DataFusion to use it (to understand what kind of changes will be needed in downstream crates). Please let me know @viirya  and @tustvold

**Review by tustvold — APPROVED** [review]
I think this is a good improvement and brings us closer to the vision of #2857. I think we should get this in, to start the process of updating the dependencies (something I fully intend to help out with), and continue to refine this in subsequent PRs.

I've taken the liberty of resolving the merge conflicts

**Comment by viirya** [comment]
Thanks @tustvold @alamb

**Comment by tustvold** [comment]
DataFusion upgrade PR - https://github.com/apache/arrow-datafusion/pull/3844

**Pull request merged**